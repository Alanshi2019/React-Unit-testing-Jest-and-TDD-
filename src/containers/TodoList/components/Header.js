import React, { Component } from "react";

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleInputKeyUp = this.handleInputKeyUp.bind(this);
  }

  render() {
    const { value } = this.state;
    return (
      <div className="header">
        <div className="header-content">
          TodoList
          <input
            className="header-input"
            placeHolder="Todo"
            data-test="input"
            value={value}
            onChange={this.handleInputChange}
            onKeyUp={this.handleInputKeyUp}
          />
        </div>
      </div>
    );
  }
  handleInputChange(e) {
    this.setState({ value: e.target.value });
  }
  handleInputKeyUp(e) {
    const { value } = this.state;
    if (e.keyCode === 13 && value) {
      this.props.addUndoItem(value);
      this.setState({
        value: ""
      });
    }
  }
}

export default Header;
