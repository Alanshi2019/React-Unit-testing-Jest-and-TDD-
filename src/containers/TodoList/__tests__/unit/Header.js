import React from "react";
import { shallow } from "enzyme";
import Header from "../../components/Header";
import { findTestWrapper } from "../../../../utils/testUtils";

it("Header render", () => {
  const wrapper = shallow(<Header />);
  expect(wrapper).toMatchSnapshot();
}); // 这个测试的目的在于 相当于是锁死了UI 只要发生一点变化 就会报错

it("Header-input check", () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, "input");
  expect(inputElem).toExist();
});

it("Header-input init", () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, "input");
  expect(inputElem.prop("value")).toEqual("");
});

it("Header-input onChange", () => {
  const wrapper = shallow(<Header />);
  const inputElem = findTestWrapper(wrapper, "input");
  //simulate可以模仿用户动作
  inputElem.simulate("change", {
    target: {
      value: "First thing I need to do"
    }
  });
  expect(wrapper.state("value")).toEqual("First thing I need to do");
});

it("Enter typed when there is nothing in the inputline.", () => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const inputElem = findTestWrapper(wrapper, "input");
  wrapper.setState({
    value: ""
  });
  //simulate可以模仿用户动作
  inputElem.simulate("keyUp", {
    keyCode: 13
  });
  expect(fn).not.toHaveBeenCalled();
});

it("Enter typed when there is something in the inputline, function called", () => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const inputElem = findTestWrapper(wrapper, "input");
  wrapper.setState({
    value: "First thing I need to do"
  });
  //simulate可以模仿用户动作
  inputElem.simulate("keyUp", {
    keyCode: 13
  });
  expect(fn).toHaveBeenCalled();
  expect(fn).toHaveBeenLastCalledWith("First thing I need to do"); //参数是....
  expect(wrapper.state("value")).toBe("");
});
