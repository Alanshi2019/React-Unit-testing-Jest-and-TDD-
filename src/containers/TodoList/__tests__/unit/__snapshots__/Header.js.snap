// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 渲染正常 1`] = `
<div
  className="header"
>
  <div
    className="header-content"
  >
    TodoList
    <input
      className="header-input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Todo"
      value=""
    />
  </div>
</div>
`;
